server/project_member.go

```diff
@@ -2,18 +2,218 @@ package server
+		repoFind := &api.RepositoryFind{ProjectID: &projectID}
+		repo, err := s.RepositoryService.FindRepository(ctx, repoFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch relevant VCS repo, Project ID: %s", c.Param("projectID"))).SetInternal(err)
+		}
+		vcsFind := &api.VCSFind{
+			ID: &repo.VCSID,
+		}
+		vcs, err := s.VCSService.FindVCS(ctx, vcsFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to find VCS for sync project member: %d", repo.VCSID)).SetInternal(err)
+		}
+		if vcs == nil {
+			return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("VCS ID not found: %d", repo.VCSID))
+		}
+		vcsProjectMemberList, err := vcsPlugin.Get(vcs.Type, vcsPlugin.ProviderConfig{Logger: s.l}).FetchRepositoryActiveMemberList(ctx,
+			common.OauthContext{
+				ClientID:     vcs.ApplicationID,
+				ClientSecret: vcs.Secret,
+				AccessToken:  repo.AccessToken,
+				RefreshToken: repo.RefreshToken,
+				Refresher:    s.refreshToken(ctx, repo.ID),
+			},
+			vcs.InstanceURL,
+			repo.ExternalID,
+		)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to fetch repository member from VCS, instance URL: %s", vcs.InstanceURL)).SetInternal(err)
+		}
+
+		// The following block will check whether the relevant principal exists in our system.
+		// If the principal does not exist, we will try to create one out of the vcs member info.
+		createList := make([]*api.ProjectMemberCreate, 0)
+		for _, projectMember := range vcsProjectMemberList {
+			if vcs.Type != projectMember.RoleProvider {
+				return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Invalid role provider, expected: %v, got: %v", vcs.Type, projectMember.RoleProvider)).SetInternal(err)
+			}
+
+			findPrincipal := &api.PrincipalFind{Email: &projectMember.Email}
+			principal, err := s.PrincipalService.FindPrincipal(ctx, findPrincipal)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal info").SetInternal(err)
+			}
+			if principal == nil { // try to create principal
+				signupInfo := &api.Signup{
+					Name:  projectMember.Name,
+					Email: projectMember.Email,
+					// Principal created via this method would have no chance to set their password.
+					// To prevent potential security issues, we use random string to set up her password.
+					// This is another safety measure since we already disallow user login via password
+					// if the principal uses external auth provider
+					Password: common.RandomString(20),
+				}
+				createdPrincipal, httpErr := trySignup(ctx, s, signupInfo, c.Get(getPrincipalIDContextKey()).(int))
+				if httpErr != nil {
+					return httpErr
+				}
+				principal = createdPrincipal
+			}
+
+			providerPayload := &api.ProjectRoleProviderPayload{
+				VCSRole:    projectMember.VCSRole,
+				LastSyncTs: time.Now().UTC().Unix(),
```

Comment:
This time should better be set before the for loop, to make sure all synced member has the exact same timestamp. Since they are all synced in one batch.